### ISSUE TITLE
ImportError: cannot import name '_psutil_linux' from partially initialized module 'psutil' (most likely due to a circular import). + "no tests ran"

### ISSUE THREAD

**Issue by Wesztman** [pytest-dev/pytest-cloud#24]
Hi! 

I wonder if you can help me with a problem I've been spending a couple of days trying to solve. 

I run `virtualenv venv && sudo venv/bin/pip install .[dev]` to create a virtual environment with all my dependencies.

I've installed pytest-cloud on both my local computer and on the remote one, as well as parallel.

I then run

`pytest tests/ --cloud-node=ubuntu@<local-ip-address> --cloud-virtualenv-path=venv/ --rsyncdir=.`

and get the following

```
Detected root dir: /workspaces/ior
Detecting connectable test nodes...
Found 1 connectable test nodes: {'ubuntu@<local-ip-address>'}
RSyncing directory structure
RSync finished
Traceback (most recent call last):
  File "/workspaces/ior/venv/bin/pytest", line 8, in <module>
    sys.exit(console_main())
  File "/workspaces/ior/venv/lib/python3.8/site-packages/_pytest/config/__init__.py", line 185, in console_main
    code = main()
  File "/workspaces/ior/venv/lib/python3.8/site-packages/_pytest/config/__init__.py", line 162, in main
    ret: Union[ExitCode, int] = config.hook.pytest_cmdline_main(
  File "/workspaces/ior/venv/lib/python3.8/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "/workspaces/ior/venv/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/workspaces/ior/venv/lib/python3.8/site-packages/pluggy/_callers.py", line 60, in _multicall
    return outcome.get_result()
  File "/workspaces/ior/venv/lib/python3.8/site-packages/pluggy/_result.py", line 60, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/workspaces/ior/venv/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "/workspaces/ior/venv/lib/python3.8/site-packages/pytest_cloud/plugin.py", line 184, in pytest_cmdline_main
    check_options(config)
  File "/workspaces/ior/venv/lib/python3.8/site-packages/pytest_cloud/plugin.py", line 409, in check_options
    node_specs = get_nodes_specs(
  File "/workspaces/ior/venv/lib/python3.8/site-packages/pytest_cloud/plugin.py", line 370, in get_nodes_specs
    caps = multi_channel.receive_each(True)
  File "/workspaces/ior/venv/lib/python3.8/site-packages/execnet/multi.py", line 259, in receive_each
    obj = ch.receive()
  File "/workspaces/ior/venv/lib/python3.8/site-packages/execnet/gateway_base.py", line 749, in receive
    raise self._getremoteerror() or EOFError()
execnet.gateway_base.RemoteError: Traceback (most recent call last):
  File "<string>", line 1088, in executetask
  File "/workspaces/ior/venv/lib/python3.8/site-packages/pytest_cloud/plugin.py", line 198, in get_node_capabilities
  File "/home/ubuntu/pytest_vscode_ior/venv/lib/python3.8/site-packages/psutil/__init__.py", line 99, in <module>
    from . import _pslinux as _psplatform
  File "/home/ubuntu/pytest_vscode_ior/venv/lib/python3.8/site-packages/psutil/_pslinux.py", line 26, in <module>
    from . import _psutil_linux as cext
ImportError: cannot import name '_psutil_linux' from partially initialized module 'psutil' (most likely due to a circular import) (/home/ubuntu/pytest_vscode_ior/venv/lib/python3.8/site-packages/psutil/__init__.py)
```

**EDIT 1:**

If I remove my psutil from my venv I get the following error instead

```
Detected root dir: /workspaces/ior
Detecting connectable test nodes...
Found 1 connectable test nodes: {'ubuntu@<local-ip-address>'}
RSyncing directory structure
RSync finished
Scheduling with 4 parallel test sessions
============================================================================================ test session starts ============================================================================================
platform linux -- Python 3.8.12, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /workspaces/ior, configfile: pyproject.toml
plugins: xdist-2.4.0, forked-1.3.0, shutil-1.7.0, cloud-5.0.3, anyio-3.4.0, virtualenv-1.7.0
<local-ip-address>_0 I / <local-ip-address>_1 I / <local-ip-address>_2 I / <local-ip-address>_3 Ibash: pytest_vscode_ior/venv/bin/python3.8: No such file or directory
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/execnet/gateway_base.py", line 997, in _send
INTERNALERROR>     message.to_io(self._io)
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/execnet/gateway_base.py", line 443, in to_io
INTERNALERROR>     io.write(header + self.data)
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/execnet/gateway_base.py", line 410, in write
INTERNALERROR>     self.outfile.flush()
INTERNALERROR> BrokenPipeError: [Errno 32] Broken pipe
INTERNALERROR> 
INTERNALERROR> During handling of the above exception, another exception occurred:
INTERNALERROR> 
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/_pytest/main.py", line 267, in wrap_session
INTERNALERROR>     config.hook.pytest_sessionstart(session=session)
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/pluggy/_hooks.py", line 265, in __call__
INTERNALERROR>     return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/pluggy/_callers.py", line 60, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/pluggy/_result.py", line 60, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/xdist/dsession.py", line 78, in pytest_sessionstart
INTERNALERROR>     nodes = self.nodemanager.setup_nodes(putevent=self.queue.put)
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/xdist/workermanage.py", line 66, in setup_nodes
INTERNALERROR>     return [self.setup_node(spec, putevent) for spec in self.specs]
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/xdist/workermanage.py", line 66, in <listcomp>
INTERNALERROR>     return [self.setup_node(spec, putevent) for spec in self.specs]
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/xdist/workermanage.py", line 69, in setup_node
INTERNALERROR>     gw = self.group.makegateway(spec)
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/execnet/multi.py", line 145, in makegateway
INTERNALERROR>     channel = gw.remote_exec(
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/execnet/gateway.py", line 132, in remote_exec
INTERNALERROR>     self._send(
INTERNALERROR>   File "/workspaces/ior/venv/lib/python3.8/site-packages/execnet/gateway_base.py", line 1003, in _send
INTERNALERROR>     raise IOError("cannot send (already closed?)")
INTERNALERROR> OSError: cannot send (already closed?)
```

**EDIT 2:**

It seems that my virtual environment python symlinks are broken on the remote after rsync when I look closer 🤔

![image](https://user-images.githubusercontent.com/54413402/144990165-e06e4e55-a7ae-4171-a299-9d0353bab8d8.png)

**EDIT 3:**

It seems that my local computer has python installed in /usr/local/bin and the remote has python in /usr/bin which seems to cause issues with the virtualenv python symlinks. Have I done anything incorrect?

**EDIT 4:**

I created a symlink on the remote with `sudo ln -s /usr/bin/python /usr/local/bin/python3` which seems to fix the above issue at least, the next error is that I get "no tests ran" 🤔

![image](https://user-images.githubusercontent.com/54413402/144995058-a1852246-8ad7-43ad-8142-c558f9e80c80.png)

**Comment by RonnyPfannschmidt** [comment]
please delete the venv and retry without sudo

**Comment by Wesztman** [comment]
@RonnyPfannschmidt 

This is what I get if I run install without sudo

```
created virtual environment CPython3.8.12.final.0-64 in 289ms
  creator CPython3Posix(dest=/workspaces/ior/venv, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/home/vscode/.local/share/virtualenv)
    added seed packages: pip==21.3.1, setuptools==58.5.3, wheel==0.37.0
  activators BashActivator,CShellActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator
Processing /workspaces/ior
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  ERROR: Command errored out with exit status 1:
   command: /workspaces/ior/venv/bin/python /workspaces/ior/venv/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py get_requires_for_build_wheel /tmp/tmpnfacsldh
       cwd: /workspaces/ior
  Complete output (3 lines):
  Warning: build in /tmp/pip-build-env-2pypk7q9/overlay/lib/python3.8/site-packages is using versioneer.py from /workspaces/ior/versioneer.py
  running egg_info
  error: [Errno 13] Permission denied
  ----------------------------------------
WARNING: Discarding file:///workspaces/ior. Command errored out with exit status 1: /workspaces/ior/venv/bin/python /workspaces/ior/venv/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py get_requires_for_build_wheel /tmp/tmpnfacsldh Check the logs for full command output.
ERROR: Command errored out with exit status 1: /workspaces/ior/venv/bin/python /workspaces/ior/venv/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py get_requires_for_build_wheel /tmp/tmpnfacsldh Check the logs for full command output.
The terminal process "bash '-c', 'virtualenv venv && venv/bin/pip install .[dev]'" terminated with exit code: 1.
```

And if i try `--user` I get

```
created virtual environment CPython3.8.12.final.0-64 in 342ms
  creator CPython3Posix(dest=/workspaces/ior/venv, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/home/vscode/.local/share/virtualenv)
    added seed packages: pip==21.3.1, setuptools==58.5.3, wheel==0.37.0
  activators BashActivator,CShellActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator
ERROR: Can not perform a '--user' install. User site-packages are not visible in this virtualenv.
The terminal process "bash '-c', 'virtualenv venv && venv/bin/pip install --user .[dev]'" terminated with exit code: 1.
```

Running with sudo was the only thing I got to work 🤔

**Comment by RonnyPfannschmidt** [comment]
im starting to get the impression that the virtualenv creation/syncing is broken in this plugin, im not familiar with the details, so i have to defer to one of its maintainers

**Comment by Wesztman** [comment]
Thanks anyway @RonnyPfannschmidt  :) my biggest issue right now is the `no tests ran`, but even though when I run only `pytest` locally it finds all  the tests. The other issues i've seemed to work around.

![image](https://user-images.githubusercontent.com/54413402/145082367-71d31a0c-545f-4375-a08b-cee7b295f6a1.png)

FYI: I also managed to install dependencies in the venv without sudo, but I still got conflicts with the psutil package on the remote side. By removing psutil from the venv I at least get to the load scheduling.

**Comment by Wesztman** [comment]
@bubenkoff and @dasm I see you're both active contributors, do you have any idea on how to fix this "no tests ran" issue? 🤔

**Comment by dasm** [comment]
@Wesztman I'm sorry it took me a while to get back to you. I haven't seen that, however it's been a long time since I used pytest-cloud.
Let me try and see what happened here.

I'll get back to you ASAP.

**Comment by dasm** [comment]
@Wesztman I don't have `python 3.8` locally, so I downloaded latest docker `python:3.8` image, but I couldn't reproduce your problem. Can you verify with docker if you're seeing the issue?

```
dasm@dell:~$ docker run -it python:3.8 bash
Unable to find image 'python:3.8' locally
[...]
Status: Downloaded newer image for python:3.8
root@209f9b461e53:/# git clone https://github.com/pytest-dev/pytest-cloud.git
Cloning into 'pytest-cloud'...
[...]
root@209f9b461e53:/# cd pytest-cloud/
root@209f9b461e53:/pytest-cloud# pip install virtualenv
[...]
root@209f9b461e53:/pytest-cloud# export USER=test
root@209f9b461e53:/pytest-cloud# make test
virtualenv .env
[...]
linters create: /pytest-cloud/.tox/linters
SKIPPED: InterpreterNotFound: python38
py38 create: /pytest-cloud/.tox/py38
py38 installdeps: -e., -r/pytest-cloud/requirements-testing.txt
py38 inst: /pytest-cloud/.tox/.tmp/package/1/pytest-cloud-5.0.3.zip
py38 installed: astroid==2.9.2,attrs==21.4.0,black==21.12b0,click==8.0.3,coverage==6.2,distlib==0.3.4,execnet==1.9.0,filelock==3.4.2,iniconfig==1.1.1,lazy-object-proxy==1.7.1,mock==4.0.3,mypy-extensions==0.4.3,packaging==21.3,pathspec==0.9.0,platformdirs==2.4.1,pluggy==1.0.0,psutil==5.9.0,py==1.11.0,pyparsing==3.0.6,pytest==6.2.5,pytest-black==0.3.12,pytest-cloud @ file:///pytest-cloud/.tox/.tmp/package/1/pytest-cloud-5.0.3.zip,pytest-forked==1.4.0,pytest-xdist==2.5.0,six==1.16.0,timeout-decorator==0.5.0,toml==0.10.2,tomli==1.2.3,typing_extensions==4.0.1,virtualenv==20.13.0,wrapt==1.13.3
py38 run-test-pre: PYTHONHASHSEED='2618860282'
py38 run-test: commands[0] | py.test tests --junitxml=/pytest-cloud/.tox/py38/log/junit-py38.xml
============================================================================================ test session starts =============================================================================================
platform linux -- Python 3.8.12, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /pytest-cloud/.tox/py38/bin/python
cachedir: .tox/py38/.pytest_cache
rootdir: /pytest-cloud, configfile: tox.ini
plugins: forked-1.4.0, cloud-5.0.3, xdist-2.5.0, black-0.3.12
collected 5 items                                                                                                                                                                                            

tests/test_plugin.py::test_schedule[1.example.com--2-100-2.example.com-user-1-200-None-None-result0] PASSED                                                                                            [ 20%]
tests/test_plugin.py::test_schedule[1.example.com--2-100-2.example.com-user-1-200-200-None-result1] PASSED                                                                                             [ 40%]
tests/test_plugin.py::test_schedule[1.example.com--2-100-2.example.com-user-1-200-None-1-result2] PASSED                                                                                               [ 60%]
tests/test_plugin.py::test_schedule[1.example.com--2-100-2.example.com-user-1-200-200-1-result3] PASSED                                                                                                [ 80%]
tests/test_plugin.py::test_schedule[1.example.com--2-100-2.example.com-user-1-200-200-1-result4] PASSED                                                                                                [100%]

----------------------------------------------------------------------- generated xml file: /pytest-cloud/.tox/py38/log/junit-py38.xml -----------------------------------------------------------------------
============================================================================================= 5 passed in 0.36s ==============================================================================================
py38-pytest-latest create: /pytest-cloud/.tox/py38-pytest-latest
SKIPPED: InterpreterNotFound: python38
__________________________________________________________________________________________________ summary ___________________________________________________________________________________________________
SKIPPED:  linters: InterpreterNotFound: python38
  py38: commands succeeded
SKIPPED:  py38-pytest-latest: InterpreterNotFound: python38
  congratulations :)
root@209f9b461e53:/pytest-cloud# 
```

**Comment by dasm** [comment]
@Wesztman I'm sorry, I misread your issue.
I confirmed the problem with `python:3.9`
```
[...]
INTERNALERROR>   File "/pytest-cloud/venv/lib/python3.9/site-packages/execnet/gateway_base.py", line 1003, in _send
INTERNALERROR>     raise IOError("cannot send (already closed?)")
INTERNALERROR> OSError: cannot send (already closed?)
```

**Comment by Wesztman** [comment]
> @Wesztman I'm sorry, I misread your issue. I confirmed the problem with `python:3.9`
> 
> ```
> [...]
> INTERNALERROR>   File "/pytest-cloud/venv/lib/python3.9/site-packages/execnet/gateway_base.py", line 1003, in _send
> INTERNALERROR>     raise IOError("cannot send (already closed?)")
> INTERNALERROR> OSError: cannot send (already closed?)
> ```

@dasm, No worries, I've made a fair bit of EDIT:s, so it could be hard to follow along ;) 

My biggest issue right now is the "no tests ran", apart from the psutil-bug which I can fix with some tinkering by removing psutil from the venv. 

The code is r-synced and I can see it on the remote, I can even SSH into the remote and activate the virtual environment and run pytest with no issues manually. 

But for some reason, there are no tests ran when going through pytest-cloud.

**Comment by Malachov** [comment]
Hello i had the same error and in my case it was because I have one venv for Windows and one venv for Linux. When I run in from Linux it breaked in Windows venv during exploring tests.

Solution for me was add to `pyproject.toml`

```toml
[tool.pytest.ini_options]
norecursedirs = [
    "venv"
]
```

**Comment by perryosa** [comment]
pip3 install --upgrade psutil
